### DISCUSSION TITLE
[v3] Load alpine plugins

### DISCUSSION THREAD

**Discussion by rahulhaque** [livewire/livewire#5986] [1. Help] [answered]
Livewire 3 by default comes with Alpine preinstalled. Alpine loads fine and I can confirm by typing `Alpine` in the browser console. However, I can't seem to load any plugins like I used to, through `resources/app.js`, as mentioned in the Alpine documentation.
```javascript
import Alpine from 'alpinejs';
import focus from '@alpinejs/focus';
import persist from '@alpinejs/persist';

window.Alpine = Alpine;
Alpine.plugin(focus);
Alpine.plugin(persist);

Alpine.store('darkMode', {
    on: Alpine.$persist(false).as('darkModeSetting'),
})

Alpine.start();
```
Did anyone else has the same issue? I was following the `$store` and `$persist` example of Alpine documentation.

**Comment by PhiloNL** [comment]
Hi @rahulhaque,

All Alpine plugins are now included by default, meaning you can remove all imports.
If you want to register custom plugins or use the Alpine object, you will have to use the ESM approach:

https://livewire.laravel.com/docs/upgrading#including-via-js-bundle

```js
import { Livewire, Alpine } from '../../vendor/livewire/livewire/dist/livewire.esm';
import customPlugin from './plugins/custom-plugin'
 
Alpine.plugin(customPlugin)

Alpine.store('darkMode', {
    on: Alpine.$persist(false).as('darkModeSetting'),
})

Livewire.start() // This will also start Alpine automatically.
```

Make sure you add the `@livewireScriptConfig` to your layout.

**Reply by rahulhaque** [reply]
Thank you @PhiloNL. I think I've missed it because there is no search feature implemented yet in the documentation. I will give it a try.

**Reply by rahulhaque** [reply]
I have followed your example as is. Disabled auto `inject_assets` in `livewire` config. Here is my layout file.
```html
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">

        <title>{{ $title ? config('app.name').' | '.$title : config('app.name') }}</title>

        @vite(['resources/css/app.css', 'resources/js/app.js'])

        @livewireStyles
    </head>
    <body>
        {{ $slot }}

        @livewireScriptConfig
    </body>
</html>
```

Here is my `app.js` file.
```javascript
import './bootstrap';

import { Livewire, Alpine } from '../../vendor/livewire/livewire/dist/livewire.esm';
 
Alpine.store('darkMode', {
    on: Alpine.$persist(false).as('darkModeSetting'),
})

Livewire.start()
```

But I am getting the following error in the console on page load.
```
Uncaught TypeError: Ka.$persist is not a function
```

**Reply by lwekuiper** [reply]
Same here. Following the documentation and provided example results in `Alpine.$persist is not a function`. 

Surrounding the logic with `document.addEventListener('alpine:init', () => { ... })` seems to work but would love to know the official solution for this.

Thanks!

**Reply by SignetPlanet** [reply]
> Hi @rahulhaque,
> 
> All Alpine plugins are now included by default, meaning you can remove all imports. If you want to register custom plugins or use the Alpine object, you will have to use the ESM approach:

It seems the `sort` plugin is not important by default. Is this plugin going away? I can't seem to import it unless I go back to the npm method but if the default doesn't include it, I suspect using npm will get me in trouble down the road as the built in defaults and npm vary over time.  Thoughts?

**Comment by rahulhaque** [comment, marked as answer]
@PhiloNL I did some more digging and this is what ultimately worked for me on default Livewire 3 setup with auto `inject_assets` on. Any other ways would show either `$persist is not a function` or the error I mentioned in my previous comment. Hope you guys will look into it. Thank you.
```javascript
document.addEventListener('alpine:init', () => {
    Alpine.store('darkMode', {
        on: Alpine.$persist(false).as('darkModeSetting'),
    });
})
```

Edit: Anyone else with the same problem? Please post your solution.

**Comment by jfp-cs** [comment]
I haven't searched other answers, so apologies if this is mentioned elsewhere.

If your persist variable is inside a module:
```javascript
// Existing (V2) code
export default {
    theme: "default",
    sidebar_open: Alpine.$persist(true).as("sidebar_open"),
...
}
```
you'll need to pass the persist value in to the module from wherever you're calling it
i.e. app.js:
```javascript
import ui from "./alpine/store/ui";
document.addEventListener("alpine:init", () => {
    Alpine.store("ui", ui({sidebar_open: Alpine.$persist(true).as("sidebar_open")}));
});
```
ui.js:
```javascript
export default ({sidebar_open}) => ({
    sidebar_open: sidebar_open,
...
})

```

I haven't extensively tested this.

**Comment by mnaderian** [comment]
https://livewire.laravel.com/docs/alpine#manually-bundling-alpine-in-your-javascript-build
